### ISSUE TITLE
Diesel doesn't see sqlite generated columns

### ISSUE THREAD

**Issue by gorbit99** [diesel-rs/diesel#2912]
## Setup
Setup a project with sqlite using the following up.sql:
```sql
create table example (
    id integer not null primary key,
    calculated integer generated always as (id*2)
);
```

### Versions

- **Rust:** 1.56.0 nightly
- **Diesel:** 1.4.7
- **Database:** Sqlite
- **Operating System** Linux

### Feature Flags

- **diesel:** sqlite

## Problem Description
Sqlite doesn't include generated columns in pragma table_info(table); only in pragma table_xinfo(table);. Therefore they do not show up in the schema.

### What are you trying to accomplish?
Use generated columns.

### Are you seeing any additional errors?
No

## Checklist

- [X] I have already looked over the [issue tracker](https://github.com/diesel-rs/diesel/issues)  and the [disussion forum](https://github.com/diesel-rs/diesel/discussions)  for similar possible closed issues.
- [X] This issue can be reproduced on Rust's stable channel. (Your issue will be
  closed if this is not the case)
- [X] This issue can be reproduced without requiring a third party crate

**Comment by weiznich** [comment]
I read a bit more about this. The [documentation] of `table_xinfo` states that it will also show hidden columns. In fact there is an additional column hidden there that contains the hidden state. Experimentation has shown that this value is set to 0 for normal columns and to 2 for generated columns. Maybe 1 means that it is an ordinary hidden column, but I was not able to check this yet. 
Additionally the documentation of [hidden columns](https://www.sqlite.org/vtab.html#hiddencol) states that hidden columns can only be created for virtual tables. Virtual tables must be registered at connection setup. This means they cannot exist in a database used to infer the schema, as diesel cli does not setup any virtual table. This means we should be fine with just using the `PRAGMA table_xinfo()` instead of `table_info` there. Any help here would be welcome.

Instructions for possible contributors:

1. Use `table_xinfo()` instead of `table_info` here: https://github.com/diesel-rs/diesel/blob/a4a70d9cdc05d722b88ea880781364098071a5e2/diesel_cli/src/infer_schema_internals/sqlite.rs#L97
2. Change the corresponding `table!` definition, so that it contains the `hidden` column here: https://github.com/diesel-rs/diesel/blob/a4a70d9cdc05d722b88ea880781364098071a5e2/diesel_cli/src/infer_schema_internals/sqlite.rs#L16-L25
3. Change the corresponding `Queryable` impl to be compatible with the changed table. You need to add an additional column of the right type here: https://github.com/diesel-rs/diesel/blob/a4a70d9cdc05d722b88ea880781364098071a5e2/diesel_cli/src/infer_schema_internals/data_structures.rs#L97-L107
4. Add the provided example as test case [here](https://github.com/diesel-rs/diesel/blob/a4a70d9cdc05d722b88ea880781364098071a5e2/diesel_cli/tests/print_schema.rs). You need to add another test function, that runs for sqlite + postgres in that file + Provide the corresponding files in [this](https://github.com/diesel-rs/diesel/tree/a4a70d9cdc05d722b88ea880781364098071a5e2/diesel_cli/tests/print_schema) folder. Look at existing test cases there for a template.

**Comment by gorbit99** [comment]
One thing, table_xinfo was added in 3.26.0, as such it's very close to the same issue [returning](https://github.com/diesel-rs/diesel/discussions/2684#discussioncomment-556697) is on.
Hidden columns on the other hand appeared in 3.4.1 (around 11 years of difference).

There is also the option of using the .schema, or the equivalent sqlite_master table (latter with a query), giving the following output given my original table with the generated column in it:
CREATE TABLE example (
    id integer primary key autoincrement not null,
    calculated generated always as (id * 3) stored
)
This solution would basically go back to way before sqlite had a version number.

Luckily these statements are highly bound by sql and definitely wouldn't need a *full* sql create statement parser for them, especially as most info would be pointless, but these include all kinds of info, such as hidden columns, generated columns, and so on. 

Issue with this would be that generated columns don't require a type as can be seen from the example, but knowing the outer functions and their return types would be sufficient in determining that.

**Comment by weiznich** [comment]
That's a good point about version compatibility. I would prefer not to have parsing code for sql statements inside of diesel. As generated columns where added with 3.31 I would just say that we check the sqlite version and if it's 3.26 or newer we use `table_xinfo()` otherwise we fall back to `table_info()`. We don't need to worry about other hidden columns here as they are a feature for virtual tables, which should not exist anyway at least not without special setup at connection creation.

**Comment by Mingun** [comment]
If I understands correctly, this issue assumes changes in the diesel_cli, which is compiled against some known version of the sqlite. So no need to check for availible features because you just can require the version that supports a necessary API.

**Comment by gorbit99** [comment]
@weiznich For now I just ignored the version check and went ahead and implemented the changes you mentioned, though they weren't complete so here's how I got it to work:

In src/infer_schema_internals/sqlite.rs changed the pragma_table_info definition to include the hidden column:
```rust
table! {
    pragma_table_info (cid) {
        cid ->Integer,
        name -> VarChar,
        type_name -> VarChar,
        notnull -> Bool,
        dflt_value -> Nullable<VarChar>,
        pk -> Bool,
        hidden -> Bool,
    }
}
```

Also added it as an unused variable to the FullTableInfo struct in the same file:
```rust
#[derive(Queryable)]
struct FullTableInfo {
    _cid: i32,
    name: String,
    _type_name: String,
    _not_null: bool,
    _dflt_value: Option<String>,
    primary_key: bool,
    _hidden: bool,
}
```

Had to change TABLE_INFO to TABLE_XINFO in two places, the one mentioned:
```rust
pub fn get_table_data(
    conn: &mut SqliteConnection,
    table: &TableName,
    column_sorting: &ColumnSorting,
) -> QueryResult<Vec<ColumnInformation>> {
    let query = format!("PRAGMA TABLE_XINFO('{}')", &table.sql_name);
    let mut result = sql::<pragma_table_info::SqlType>(&query).load(conn)?;
    match column_sorting {
        ColumnSorting::OrdinalPosition => {}
        ColumnSorting::Name => {
            result.sort_by(|a: &ColumnInformation, b: &ColumnInformation| {
                a.column_name.partial_cmp(&b.column_name).unwrap()
            });
        }
    };
    Ok(result)
}
```

And where the primary keys are determined:
```rust
pub fn get_primary_keys(
    conn: &mut SqliteConnection,
    table: &TableName,
) -> QueryResult<Vec<String>> {
    let query = format!("PRAGMA TABLE_XINFO('{}')", &table.sql_name);
    let results = sql::<pragma_table_info::SqlType>(&query)
        .load::<FullTableInfo>(conn)?;
    Ok(results
        .into_iter()
        .filter_map(|i| if i.primary_key { Some(i.name) } else { None })
        .collect())
}
```

From line 97 in src/infer_schema_internals/data_structures.rs: (Added bool to the type and the where statement)
```rust
#[cfg(feature = "sqlite")]
impl<ST> Queryable<ST, Sqlite> for ColumnInformation
where
    (i32, String, String, bool, Option<String>, bool, bool):
        FromStaticSqlRow<ST, Sqlite>,
{
    type Row = (i32, String, String, bool, Option<String>, bool, bool);

    fn build(row: Self::Row) -> deserialize::Result<Self> {
        Ok(ColumnInformation::new(row.1, row.2, None, !row.3))
    }
}
```

Now the cli builds, and errors out with the error: "The type 'generated always' isn't supported', because the output of table_xinfo looks like this:
```
0|id|integer|1||1|0
1|calculated|generated always|0||0|3
```
Notice the 3 in the output? Fun eh?  Checking the source of sqlite, there's a nice little comment here: https://github.com/sqlite/sqlite/blob/91f621531dc1cb9ba5f6a47eb51b1de9ed8bdd07/src/pragma.c#L1165
Saying 2 -> virtual generated, 3 -> stored generated (and 1 is indeed just hidden)

The returned type seems to always contain whatever was put between the column name and 'as', so possible values:
empty string
<type>
<type> generated always
generated always

This makes it especially complex to find out the type.

**Comment by weiznich** [comment]
Thanks for the great summary :+1: 
It would be great if you can open a WIP PR containing this changes. This would make it much easier to comment on certain parts of your solution.
I would change the `hidden` column to `Integer` as it can contain more values than only 0 and 1. 

> The returned type seems to always contain whatever was put between the column name and 'as', so possible values:
> empty string
> _type_
> _type_ generated always
> generated always
>
> This makes it especially complex to find out the type.

As for now I would like to keep this really simple. Rejecting generated columns without explicit type should be fine for the first step, as there is always the option to specify the type there as you've done in your example case. We can improve on this later on, but I do not really see a good solution there. At least not without having to parse the actual generate expression and trying to figure out the type from there. What could be easily done is to just emit a better error message in this case pointing out that diesel cannot infer the correct type in this case.

**Comment by gorbit99** [comment]
> We can improve on this later on, but I do not really see a good solution there.
Btw, I made a post on the sqlite forum about it, maybe there's a good solution, but we'll see.

**Comment by weiznich** [comment]
:+1: 
Can you link the corresponding post here for future reference?

**Comment by gorbit99** [comment]
https://sqlite.org/forum/forumpost/8ebb993012

**Comment by gorbit99** [comment]
Working on this issue again, because I found more time all of a sudden.

So first and foremost, after reading the sqlite docs more carefully, this:

> The datatype and collating sequence of the generated column are determined only by the datatype and COLLATE clause on the column definition. The datatype and collating sequence of the GENERATED ALWAYS AS expression have no affect on the datatype and collating sequence of the column itself.

To me seems to mean, that if the type isn't specified, then that means, like in the rest of sqlite, it will take the blob type. So that's one issue out of the picture.

However, when I mentioned, that "generated always" ate itself into the table_xinfo description... that seems to have been a bug, as now it isn't. This was fixed in 3.37.0, so from 3.31 to 3.36, this issue existed. About 2 years.

**Comment by gorbit99** [comment]
> Working on this issue again, because I found more time all of a sudden.
> 
> So first and foremost, after reading the sqlite docs more carefully, this:
> 
> > The datatype and collating sequence of the generated column are determined only by the datatype and COLLATE clause on the column definition. The datatype and collating sequence of the GENERATED ALWAYS AS expression have no affect on the datatype and collating sequence of the column itself.
> 
> To me seems to mean, that if the type isn't specified, then that means, like in the rest of sqlite, it will take the blob type. So that's one issue out of the picture.
> 
> However, when I mentioned, that "generated always" ate itself into the table_xinfo description... that seems to have been a bug, as now it isn't. This was fixed in 3.37.0, so from 3.31 to 3.36, this issue existed. About 2 years.

Ok, so cli has sqlite bundled into it, which means theoretically the second part of this message shouldn't be important and it should be ok to leave it like that, though I'll let the smarter people decide

**Comment by weiznich** [comment]
Closed by #2916
